Documentation: Sequences (#791)
ocaml/ocaml.org: data/tutorials/ds_05_seq.md

description: >
  Learn about an OCaml's most-used, built-in data types
category: "data-structures"
date: 2023-01-12T09:00:00-01:00
---

# Sequences

## Prerequisites

| Concept | Status | Documentation | Reference |
|---|---|---|---|
| Basic types | Mandatory | | |
| Functions | Mandatory | | |
| Lists   | Mandatory |   |   |
| Options | Recommended |   |   |
| Arrays  | Nice to have |   |   |

## Introduction

Sequences look a lot like lists. However from a pragmatic perspective, one
should imagine they may be infinite. That's the key intuition to understanding
and using sequences.

One way to look at a value of type `'a Seq.t` is to consider it as an icicle, a
frozen stream of data. To understand this analogy, consider how sequences are
defined in the standard library:
```ocaml
type 'a node =
  | Nil
  | Cons of 'a * 'a t
and 'a t = unit -> 'a node
```
This is the mutually recursive definition of two types; `Seq.node` which is
almost the same as `list`:
```ocaml
type 'a list =
  | []
  | (::) of 'a * 'a list
```
and `Seq.t` which is merely a type alias for `unit -> 'a Seq.node`. The whole
point of this definition is the type of the second argument of `Seq.Cons`, which

> is a function returning a sequence while its `list` sibling is a list. Let's

Reviewer: ```suggestion
is a function returning a sequence while its `list` counterpart returns a list. Let's
```

In OCaml, any value `a` of type `t` can be turned into a constant function by
writing `fun _ -> a`, which has type `'a -> t`. When writing `fun () -> a`
instead, we get a function of type `unit -> t`. Such a function is called a
[_thunk_](https://en.wikipedia.org/wiki/Thunk). Using this terminology, sequence
values are thunks. With the analogy used earlier, `a` is frozen in its thunk.

Here is how to build seemingly infinite sequences of integers:
```ocaml
# let rec ints n : int Seq.t = fun () -> Seq.Cons (n, ints (n + 1))
val ints : int -> int Seq.t = <fun>
```
The function `ints n` look as if building the infinite sequence
$(n; n + 1; n + 2; n + 3;...)$. In reality, since there isn't an infinite
amount of distinct values of type `int`, those sequences are not increasing,
when reaching `max_int` the values will circle down to `min_int`. They are
ultimately periodic.

The OCaml standard library contains a module on sequences called
[`Seq`](/releases/5.0/api/Seq.html). It contains a `Seq.iter` function, which
has the same behaviour as `List.iter`. Writing this:
```ocaml
# Seq.iter print_int (ints 0);;
```
in an OCaml top-level means: “print integers forever” and you have to type

> `Crtl-C` to interrupt the execution. Perhaps more interestingly, the following

Reviewer: ```suggestion
`Ctrl-C` to interrupt the execution. Perhaps more interestingly, the following
```

"README.md" |> open_in |> Seq.unfold input_line_opt |> Seq.iter print_endline
```

Although this can be an appealing style, bear in mind it does not prevent from
taking care of open files. While the code above is fine, this one no longer is:
```ocaml
"README.md" |> open_in |> Seq.unfold input_line_opt |> Seq.take 10 |> Seq.iter print_endline
```
Here, `close_in` will never be called over the input channel opened on `README.md`.

## Sequences for Conversions

Throughout the standard library, sequences are used as a bridge to perform
conversions between many datatypes. For instance, here are the signatures of
some of those functions:
* Lists
  ```ocaml
  val List.of_seq : 'a list -> 'a Seq.t
  val List.to_seq : 'a Seq.t -> 'a list
  ```
* Arrays
  ```ocaml
  val Array.of_seq : 'a array -> 'a Seq.t
  val Array.to_seq : 'a Seq.t -> 'a array
  ```
* Strings
  ```ocaml
  val String.of_seq : string -> char Seq.t
  val String.to_seq : char Seq.t -> string
  ```
Similar functions are also provided for sets, maps, hash tables (`Hashtbl`) and
others (except `Seq`, obviously). When implementing a datatype module, it is
advised to expose `to_seq` and `of_seq` functions.

## Miscellaneous

> There are a couple of related Libraries, all providing means to handle large

Reviewer: ```suggestion
There are a couple of related libraries, all providing means to handle large
```
